src/crimson/os/seastore/seastore_types.h

```diff
@@ -36,28 +36,148 @@ struct seastore_meta_t {
   }
 };
 
+// internal segment id type of segment_id_t, basically
+// this is a unsigned int with the top "device_id_len"
+// bits representing the id of the device on which the
+// segment resides
+using internal_segment_id_t = uint32_t;
+// type of device id
+using device_id_t = uint8_t;
+// order of device_id_t
+constexpr uint16_t device_id_len = 4;
+// maximum devices supported
+constexpr uint16_t max_devices = 1 << device_id_len;
+
+// segment ids without a device id encapsulated
+using device_segment_id_t = internal_segment_id_t;
+
+// mask for segment manager id
+constexpr internal_segment_id_t SM_ID_MASK =
+  0xF << (std::numeric_limits<internal_segment_id_t>::digits - device_id_len);
+// default internal segment id
+constexpr internal_segment_id_t DEFAULT_INTERNAL_SEG_ID =
+  std::numeric_limits<internal_segment_id_t>::max() - 1;
+
 // Identifies segment location on disk, see SegmentManager,
-using segment_id_t = uint32_t;
+struct segment_id_t {
+  internal_segment_id_t segment = DEFAULT_INTERNAL_SEG_ID;
```

Comment:
does the internal_segment_id_t always include device_id_t in the high 4 bits?

Reply:
what is the benefit to use a combined integer to save device_id_t and device_segment_id_t, every place when to use them need calculate.  how about use a structure with two members?

Reply:
This is fine.  Otherwise, you'd need to use a bitfield.